conda/notices/cache.py

```diff
@@ -0,0 +1,150 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2012 Anaconda, Inc
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+Handles all caching logic including:
+  - Retrieving from cache
+  - Saving to cache
+  - Determining whether not certain items have expired and need to be refreshed
+"""
+import json
+import logging
+import os
+from datetime import datetime, timezone
+from functools import wraps
+from pathlib import Path
+from typing import Optional, Sequence, Set
+
+from conda._vendor.appdirs import user_cache_dir
+from conda.base.constants import APP_NAME, NOTICES_CACHE_SUBDIR, NOTICES_CACHE_FN
+from conda.utils import ensure_dir_exists, safe_open
+
+from .types import ChannelNoticeResponse, ChannelNotice
+
+logger = logging.getLogger(__name__)
+
+
+def cached_response(func):
+    @wraps(func)
+    def wrapper(url: str, name: str):
+        cache_dir = get_notices_cache_dir()
+        cache_val = get_notice_response_from_cache(url, name, cache_dir)
+
+        if cache_val:
+            return cache_val
+
+        return_value = func(url, name)
+        if return_value is not None:
+            write_notice_response_to_cache(return_value, cache_dir)
+
+        return return_value
+
+    return wrapper
+
+
+def is_notice_response_cache_expired(channel_notice_response: ChannelNoticeResponse) -> bool:
+    """
+    This checks the contents of the cache response to see if it is expired.
+
+    If for whatever reason we encounter an exception while parsing the individual
+    messages, we assume an invalid cache and return true.
+    """
+    now = datetime.now(timezone.utc)
+
+    def is_channel_notice_expired(expired_at: Optional[datetime]) -> bool:
+        """
+        If there is no "expired_at" field present assume it is expired
+        """
+        if expired_at is None:
+            return True
+        return expired_at < now
+
+    return any(
+        (is_channel_notice_expired(chn.expired_at) for chn in channel_notice_response.notices)
+    )
+
+
+@ensure_dir_exists
+def get_notices_cache_dir() -> Path:
+    """Returns the location of the notices cache directory as a Path object"""
+    cache_dir = user_cache_dir(APP_NAME, appauthor=APP_NAME)
+
+    return Path(cache_dir).joinpath(NOTICES_CACHE_SUBDIR)
+
+
+def get_notices_cache_file() -> Path:
+    """Returns the location of the notices cache file as a Path object"""
+    cache_dir = get_notices_cache_dir()
+    cache_file = cache_dir.joinpath(NOTICES_CACHE_FN)
+
+    if not cache_file.is_file():
+        cache_file.write_text("")
```

Comment:
Good catch! I actually adding a context manager specifically for this scenario called `safe_open` which provides a meaningful error message instead of blowing up as a big stacktrace.